plugin/sampling/strategystore/adaptive/processor.go

```diff
@@ -0,0 +1,506 @@
+	// is stored at the head of the slice.
+	throughputs []*throughputBucket
+
+	// strategyResponses contains the sampling strategies for every service.
+	strategyResponses map[string]*sampling.SamplingStrategyResponse
+
+	logger *zap.Logger
+
+	weightVectorCache *weightVectorCache
+
+	probabilityCalculator calculationstrategy.ProbabilityCalculator
+
+	// followerProbabilityInterval determines how often the follower processor updates its probabilities.
+	// Given only the leader writes probabilities, the followers need to fetch the probabilities into
+	// cache.
+	followerProbabilityInterval time.Duration
+
+	serviceCache []samplingCache
+
+	operationsCalculatedGauge     metrics.Gauge
+	calculateProbabilitiesLatency metrics.Timer
+}
+
+// NewProcessor creates a new sampling processor that generates sampling rates for service operations
+func NewProcessor(
+	opts Options,
+	hostname string,
+	storage samplingstore.Store,
+	electionParticipant leaderelection.ElectionParticipant,
+	metricsFactory metrics.Factory,
+	logger *zap.Logger,
+) (ss.StrategyStore, error) {
+	if opts.CalculationInterval == 0 || opts.AggregationBuckets == 0 {
+		return nil, errNonZero
+	}
+	if opts.BucketsForCalculation < 1 {
+		return nil, errBucketsForCalculation
+	}
+	metricsFactory = metricsFactory.Namespace(metrics.NSOptions{Name: "adaptive_sampling_processor"})
+	return &processor{
+		Options:             opts,
+		storage:             storage,
+		probabilities:       make(model.ServiceOperationProbabilities),
+		qps:                 make(model.ServiceOperationQPS),
+		hostname:            hostname,
+		strategyResponses:   make(map[string]*sampling.SamplingStrategyResponse),
+		logger:              logger,
+		electionParticipant: electionParticipant,
+		// TODO make weightsCache and probabilityCalculator configurable
+		weightVectorCache:             newWeightVectorCache(),
+		probabilityCalculator:         calculationstrategy.NewPercentageIncreaseCappedCalculator(1.0),
+		followerProbabilityInterval:   defaultFollowerProbabilityInterval,
+		serviceCache:                  []samplingCache{},
+		operationsCalculatedGauge:     metricsFactory.Gauge(metrics.Options{Name: "operations_calculated"}),
+		calculateProbabilitiesLatency: metricsFactory.Timer(metrics.TimerOptions{Name: "calculate_probabilities"}),
+	}, nil
+}
+
+func (p *processor) GetSamplingStrategy(service string) (*sampling.SamplingStrategyResponse, error) {
+	p.RLock()
+	defer p.RUnlock()
+	if strategy, ok := p.strategyResponses[service]; ok {
+		return strategy, nil
+	}
+	return p.generateDefaultSamplingStrategyResponse(), nil
+}
+
+// Start initializes and starts the sampling processor which regularly calculates sampling probabilities.
+func (p *processor) Start() error {
+	p.logger.Info("starting sampling processor")
+	p.shutdown = make(chan struct{})
+	if starter, ok := p.electionParticipant.(jio.Starter); ok {
+		starter.Start()
+	}
+	p.loadProbabilities()
```

Comment:
- master: since there's no probability to compare against to determine if an operation is using adaptive sampling, the processor will start calculating probabilities again from the default prob 0.001.
- follower: will return default probabilities to agents/clients.